rule all:
    input:
        multiext("results/genome/bn.bnir.zs11.v0.genome.fa", ".amb", ".ann", ".bwt", ".pac", ".sa"),
        "results/custom_bed/bn.ZS11.v0.transcript.bed", "results/custom_bed/bn.ZS11.v0.gene.bed",
        "results/gene_info.tsv.gz",
        json = "results/igv_genome/bn.bnir.zs11.v0.genome.json"


rule download_genome:
    output:
        "rawdata/Brassica_napus.ZS11.v0.genome.fa.gz"
    params:
        url = "https://yanglab.hzau.edu.cn/static/bnir/assets//genomic_sequence/BnIRData/AACC.Brassica_napus/ZS11/v0/Brassica_napus.ZS11.v0.genome.fa.gz"
    shell:
        '''
        curl -L {params.url} > {output}
        '''


rule download_gff:
    output:
        "rawdata/Brassica_napus.ZS11.v0.gene.gff3.gz"
    params:
        url = "https://yanglab.hzau.edu.cn/static/bnir/assets//genomic_sequence/BnIRData/AACC.Brassica_napus/ZS11/v0/Brassica_napus.ZS11.v0.gene.gff3.gz"
    shell:
        '''
        curl -L {params.url} > {output}
        '''


rule genome_index:
    input:
        "rawdata/Brassica_napus.ZS11.v0.genome.fa.gz"
    output:
        fa = "results/genome/bn.bnir.zs11.v0.genome.fa",
        fai = "results/genome/bn.bnir.zs11.v0.genome.fa.fai",
        dict = "results/genome/bn.bnir.zs11.v0.genome.dict"
    shell:
        '''
        gzip -dc {input} > {output.fa}
        samtools faidx {output.fa}
        gatk CreateSequenceDictionary -R {output.fa}
        '''


rule bwa_index:
    input:
        "results/genome/bn.bnir.zs11.v0.genome.fa"
    output:
        multiext("results/genome/bn.bnir.zs11.v0.genome.fa", ".amb", ".ann", ".bwt", ".pac", ".sa")
    shell:
        "bwa index {input}"


rule make_gtf:
    input:
        "rawdata/Brassica_napus.ZS11.v0.gene.gff3.gz"
    output:
        "results/gtf/bn.ZS11.v0.make.gtf"
    script:
        "scripts/make_gtf.R"


rule bgzip:
    input:
        "{prefix}"
    output:
        "{prefix}.gz"
    shell:
        "bgzip {input} > {output}"


rule tabix_index:
    input:
        "{prefix}.gz"
    output:
        "{prefix}.gz.tbi"
    shell:
        "tabix {input}"


rule custom_bed:
    input:
        gtf = "results/gtf/bn.ZS11.v0.make.gtf"
    output:
        transcript = "results/custom_bed/bn.ZS11.v0.transcript.bed",
        gene = "results/custom_bed/bn.ZS11.v0.gene.bed"
    script:
        "scripts/custom_bed.R"


rule scrape_annotation:
    output:
        protected("results/bnir_scrape_annotation.csv")
    script:
        "scripts/scrape_annotation.py"


rule clean_gene_info:
    input:
        "results/bnir_scrape_annotation.csv"
    output:
        "results/gene_info.tsv.gz"
    script:
        "scripts/clean_gene_info.R"


rule make_gff:
    input:
        "rawdata/Brassica_napus.ZS11.v0.gene.gff3.gz"
    output:
        "results/make_gff/bn.bnir.zs11.v0.gene.gff"
    script:
        "scripts/make_gff.R"


rule make_igv_genome:
    input:
        fa = "results/genome/bn.bnir.zs11.v0.genome.fa",       # IGV can only read ungzipped FASTA
        fai = "results/genome/bn.bnir.zs11.v0.genome.fa.fai",
        gff = "results/make_gff/bn.bnir.zs11.v0.gene.gff"           # IGV's search only function with ungzipped GFF/GTF
    output:
        outdir = directory("results/igv_genome"),
        json = "results/igv_genome/bn.bnir.zs11.v0.genome.json"
    params:
        id =  "ZS11-v0",
        name = "B. napus (ZS11-v0)",
#         chromosome_order = ["chr01", "chr02", "chr03", "chr04", "chr05", "chr06", "chr07", "chr08", "chr09", "chr10", "chr11", "chr12"]
    script:
        "scripts/make_igv_genome.py"
